### DISCUSSION TITLE
Server-Sent Events being blocked on Windows?

### DISCUSSION THREAD

**Discussion by jimafisk** [pocketbase/pocketbase#4539] [Q&A]
We already have a work around for this issue, so no action is needed here, but I thought I'd start a discussion in case other folks are running into a similar issue.

**Scenario:**
We were using an OAuth2 login (Gitea) using Server-Sent Events (SSE) and it worked great for most users, but certain Windows users could not log in - they'd just see a blank popup window. We had difficulty replicating the issue with other Windows machines / VMs but 2 users from different companies raised the issue to us. In both cases, the user likely had a corporate Active Directory that was implementing security policies on their machines and potentially blocking requests.

**Console Error:**
I used `err.originalError` per Gani's suggestion (https://github.com/pocketbase/js-sdk/issues/205#issuecomment-1584721125) and found the issue was:
```
Error: EventSource connect took too long.
```
**Work Around:**
Switching from SSE to manual code exchange seemed to resolve the issue for us.

<hr>

Possibly related discussion: https://github.com/pocketbase/pocketbase/discussions/2958

**Comment by ganigeorgiev** [comment]
`EventSource connect took too long` is usually returned when for some reason the client cannot establish connection with the PocketBase server.

While the error message could be the same, I'm not sure that the linked discussion is directly related to your issue (_as explained in https://github.com/pocketbase/pocketbase/discussions/2958#discussioncomment-6501454, the issue there is most likely because of the React Native flipper dev plugin_).

SSE in practice is no different than a regular HTTP GET connection (there is no hacky protocol upgrade step like in WebSocket) and it shouldn't trigger firewall/security policies. I could imagine a policy for how long a connection can remain open but in your case it seems that the connection was never established in the first place.

Without a code sample showing how the client was initialized or what is the environment setup (is PocketBase behind a proxy, inside Docker, etc.) would be difficult to pinpoint the exact reason for the error.

If the above is running locally and you use `http://localhost:8090` as base url for the SDK it is possible to cause the above error depending on the platform hosts configuration and how the address will be resolved (_whether it will try first/only with IPv6 or the IPv4 version_) since by default we bind only to the IPv4 version `127.0.0.1:8090` (see also https://github.com/pocketbase/js-sdk/issues/21). But if that was the case then other regular browser requests to the PocketBase APIs would also fail, so I'm not sure.

Another possible reason since the above is executed in a browser, if the PocketBase server is running locally and there were multiple browser tabs with active realtime connections, then the above error could be because of the browsers SSE HTTP/1 ~6 connections limit (_see the warning in https://developer.mozilla.org/en-US/docs/Web/API/EventSource; this is not an issue on HTTP/2 which the Go std server automatically uses when started with https_).

**Reply by jimafisk** [reply]
@ganigeorgiev thank you for being so quick to respond as always! I don't want to waste your time debugging this on our behalf as we're happy with our workaround, just wanted to put something out there in case someone is having similar issues. I'll put a few more comments here, but no need to respond!

> EventSource connect took too long is usually returned when for some reason the client cannot establish connection with the PocketBase server.

The users reporting the issues were getting CORS errors in the console. We were confused as to why this was happening for them but not other users:
```
Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://ourpocketbaseserver.com/api/realtime. (Reason: CORS request did not succeed). Status code: (null).
```

> is PocketBase behind a proxy, inside Docker, etc.

PocketBase is not behind a proxy and is not in a docker container. We just put the compiled binary on a server and use systemd to run it.

> if the PocketBase server is running locally and there were multiple browser tabs with active realtime connections, then the above error could be because of the browsers SSE HTTP/1 ~6 connections limit

We weren't able to recreate the issue with a local PocketBase so we were using a remote/deployed instance. We had the 2 users close all tabs, remove all browser plugins, and open incognito / private browsing windows but all tests resulted in the same behavior for those users only. Mac users from the same org could login without issues and our dev team uses Linux and did not have issues logging in. We also tested with Windows computers we own and virtual windows computers and could not replicate the issue (login worked fine).

<hr> 

The main takeaway was that we ended up using `authWithOAuth2Code` directly instead of `authWithOAuth2` which apparently calls the former under the hood (https://github.com/pocketbase/pocketbase/discussions/2906#discussioncomment-6468116)

Changing from something like:
```js
const authData = await $client.collection("staff").authWithOAuth2({
  provider: "gitea",
  urlCallback: url => {
    window.open(url, "_blank", "");
  }
});
```

To something like this worked for us:
```js
const authData = await $client.collection("staff").authWithOAuth2Code(
  provider.name,
  params.get("code"),
  provider.codeVerifier,
  env.cms.redirectUrl,
);
```
